src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -51,78 +50,90 @@ export class BrowserAdapterEventManager {
     };
 
     private processDeferredEvents(): void {
-        this.deferredEvents = this.deferredEvents.filter(
-            event => this.processEvent(event.eventType, event.eventArgs, true) === false,
-        );
+        const stillDeferred: DeferredEventDetails[] = [];
+        this.deferredEvents.forEach(deferredEvent => {
+            if (this.eventsToApplicationListenersMapping[deferredEvent.eventType]) {
+                try {
+                    this.processEvent(deferredEvent.eventType, deferredEvent.eventArgs);
+                    deferredEvent.deferredResolution.resolveHook(true);
+                } catch (error) {
+                    console.error(error);
+                    deferredEvent.deferredResolution.rejectHook();
+                }
+            } else {
+                stillDeferred.push(deferredEvent);
+            }
+        });
+        this.deferredEvents = stillDeferred;
     }
 
-    private async forwardEventToApplicationListener(
+    private forwardEventToApplicationListener(
         listener: ApplicationListener,
         eventArgs: any[],
     ): Promise<any> {
-        const result = listener(...eventArgs);
+        let result: any;
+        try {
+            result = listener(...eventArgs);
+        } catch (error) {
+            console.error('Error thrown in application listener: ', error);
+            throw error;
+        }
         if (!!result && typeof result.then === 'function') {
             //wrapping application listener promise responses in a 4-minute promise race
             //prevents the service worker going idle before a response is sent
-            await this.promiseFactory.timeout(result, FOUR_MINUTES);
+            return this.promiseFactory.timeout(result, FOUR_MINUTES);
         } else {
```

Comment:
Do we have any cases where an application handler will be returning a non-undefined, non-promise result? I'm not sure the "delay for 2 minutes" approach would be a good idea in that case, we probably don't want to delay actual responses for 2 minutes. I don't think we have a great solution for that case - we don't have a way to simultaneously report a response back and also indicate the 2 minute deferal.

I think we should probably fail fast and log an error if that happens, and update any such cases that occur in practice by auditing them for any fire and forget behavior, updating them to ensure any such behavior is captured within a Promise, and then returning a Promise with the original result.